snuba/subscriptions/scheduler_consumer.py

```diff
@@ -0,0 +1,306 @@
+from __future__ import annotations
+
+import logging
+from datetime import datetime, timedelta
+from typing import Callable, Mapping, MutableMapping, NamedTuple, Optional, Sequence
+
+from arroyo import Message, Partition, Topic
+from arroyo.backends.abstract import Consumer
+from arroyo.backends.kafka import KafkaConsumer, KafkaPayload
+from arroyo.processing import StreamProcessor
+from arroyo.processing.strategies import ProcessingStrategy
+from arroyo.processing.strategies.abstract import ProcessingStrategyFactory
+from arroyo.synchronized import commit_codec
+from arroyo.types import Position
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.utils.metrics import MetricsBackend
+from snuba.utils.streams.configuration_builder import build_kafka_consumer_configuration
+from snuba.utils.types import Interval, InvalidRangeError
+
+logger = logging.getLogger(__name__)
+
+
+class MessageDetails(NamedTuple):
+    offset: int
+    orig_message_ts: datetime
+
+
+class Tick(NamedTuple):
+    partition: int
+    offsets: Interval[int]
+    timestamps: Interval[datetime]
+
+    def time_shift(self, delta: timedelta) -> Tick:
+        """
+        Returns a new ``Tick`` instance that has had the bounds of its time
+        interval shifted by the provided delta.
+        """
+        return Tick(
+            self.partition,
+            self.offsets,
+            Interval(self.timestamps.lower + delta, self.timestamps.upper + delta),
+        )
+
+
+class CommitLogTickConsumer(Consumer[Tick]):
+    """
```

Comment:
I think we should add notes about the fact that this consumer now generates the tick from the metadata in the payload of the commit log and does not uses partition offset and timestamp from the main topic it consumes.